xapian-core/weight/tfidfweight.cc

```diff
@@ -353,12 +355,77 @@ TfIdfWeight::get_wtn(double wt, wt_norm wt_normalization) const
     return wt;
 }
 
+static inline void
+parameter_error(const char* message)
+{
+    Xapian::Weight::Internal::parameter_error(message, "tfidf");
+}
+
 TfIdfWeight *
 TfIdfWeight::create_from_parameters(const char * p) const
 {
     if (*p == '\0')
 	return new Xapian::TfIdfWeight();
-    return new Xapian::TfIdfWeight(p);
+    string wdf_normalisation, idf_normalisation, wt_normalisation;
+    wdf_norm wdf_normalisation_ = wdf_norm::NONE;
+    idf_norm idf_normalisation_ = idf_norm::TFIDF;
+    wt_norm wt_normalisation_ = wt_norm::NONE;
+    if (!Xapian::Weight::Internal::param_name(&p, wdf_normalisation))
+	parameter_error("Parameter 1 (wdf_normalisation) is invalid");
+    if (!Xapian::Weight::Internal::param_name(&p, idf_normalisation))
+	parameter_error("Parameter 2 (idf_normalisation) is invalid");
+    if (!Xapian::Weight::Internal::param_name(&p, wt_normalisation))
+	parameter_error("Parameter 3 (wt_normalisation) is invalid");
+    if (*p)
+	parameter_error("Extra data after wt_normalisation");
+    map<string, wdf_norm> wdfnorm {
+	{ "NONE", wdf_norm::NONE },
+	{ "BOOLEAN", wdf_norm::BOOLEAN },
+	{ "SQUARE", wdf_norm::SQUARE },
+	{ "LOG", wdf_norm::LOG },
+	{ "PIVOTED", wdf_norm::PIVOTED },
+	{ "LOG_AVERAGE", wdf_norm::LOG_AVERAGE },
+	{ "AUG_LOG", wdf_norm::AUG_LOG },
+	{ "SQRT", wdf_norm::SQRT },
+	{ "AUG_AVERAGE", wdf_norm::AUG_AVERAGE },
+	{ "MAX", wdf_norm::MAX },
+	{ "AUG", wdf_norm::AUG }
+    };
+    map<string, idf_norm> idfnorm {
+	{ "NONE", idf_norm::NONE },
+	{ "TFIDF", idf_norm::TFIDF },
+	{ "SQUARE", idf_norm::SQUARE },
+	{ "FREQ", idf_norm::FREQ },
+	{ "PROB", idf_norm::PROB },
+	{ "PIVOTED", idf_norm::PIVOTED },
+	{ "GLOBAL_FREQ", idf_norm::GLOBAL_FREQ },
+	{ "LOG_GLOBAL_FREQ", idf_norm::LOG_GLOBAL_FREQ },
+	{ "INCREMENTED_GLOBAL_FREQ", idf_norm::INCREMENTED_GLOBAL_FREQ },
+	{ "SQRT_GLOBAL_FREQ", idf_norm::SQRT_GLOBAL_FREQ }
+    };
+    map<string, wt_norm> wtnorm {
+	{ "NONE", wt_norm::NONE }
+    };
```

Comment:
This will have to construct these maps each time this method is called and then does one lookup in each - in general it very rarely makes sense to construct a complex data structure at run-time for a single lookup.  Building the data structure inevitably has to make a pass over all the data being put in it making in at least O(n), which is what a simple linear search over the data would be.

Constructing a map is O(n*log(n)) if the input data isn't sorted, as here (pre-sorting the input would help as then it's O(n)) and then looking up in the map is O(log(n)) - if you just did a simple `if`-chain that'd be O(n).  If the input data is sorted, binary chop is O(log(n)) (and `std::lower_bound()` provides a binary chop implementation) though that does rely on people adding new entries actually maintaining the sorted order.

Admittedly big-O scaling isn't necessarily the key concern when you have a small number of entries because the scale factor can matter more - e.g. a fast O(n²) can beat a slow O(log(n)) for bounded n - but I'd expect the scale factor to be worse for building a std::map keyed on strings.

Another issue this hits is that the initialiser is probably essentially an array with string constants in it, and in a shared library each string constant in an array needs relocating by the dynamic linker at library load time, which makes loading the library slower so we try to avoid these.  If the string constants are sensibly bounded in length (as here) there's a trick to avoid this problem - however we have a mechanism to build lookup tables for string to enum (or integer) mapping at compile time which I think would make sense to also use here.

It's a perl module - you write a script which instantiates an object `$hdr` and then for each (string, enum) pair calls:

```
$hdr->add($string, $enum);
```

This generates a header file which you then include from the code and to look up string `s`:

```
int r = keyword(whatever_the_generated_table_is_called, s.data(), s.size())
if (r < 0) {
    // Handle an unknown token.
    throw some_exception();
}
// Known token, so safe to cast to enum_type.
auto code = static_cast<enum_type>(r);
```

(There's also `keyword2()` which handles lookup tables which are large enough to need 2-byte offsets.)

This is used for keyword lookups in a number of places already - e.g. for stemmer names see `xapian-core/languages/collate-sbl` which generates `xapian-core/languages/sbl-dispatch.h` and is used in `xapian-core/languages/stem.cc`.  This script also has to allocate enum values and create an enum definition, which you don't need here as they are already defined so you don't need that part.

You'll want a separate lookup table for each type (or maybe just hard-code a simple `== "NONE"` check for the last one for now.

The way the lookup works is essentially to dispatch first on the length of the input via a lookup table (which is O(1)) and then perform a binary chop just within a list of known tokens with that length, so it should be more efficient that a binary chop on the whole list and the lookup tables are compact.